File: 1. 단계별로_풀어보기/30~39/33. 트리/4. 1991.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
childArr = list([None, None] for i in range(256))

def VLR(now):
  result = chr(now)
  lc, rc = childArr[now]

  if lc:
    result += VLR(lc)
  if rc:
    result += VLR(rc)
  return result

def LVR(now):
  result = chr(now)
  lc, rc = childArr[now]

  if lc:
    result = LVR(lc) + result
  if rc:
    result += LVR(rc)
  return result

def LRV(now):
  result = ''
  lc, rc = childArr[now]

  if lc:
    result += LRV(lc)
  if rc:
    result += LRV(rc)
  
  result += chr(now)
  return result

asciiA = ord('A')
asciiZ = ord('Z')

for _ in range(N):
  parent, lc, rc = input().split()

  if asciiA <= ord(lc) <= asciiZ:
    childArr[ord(parent)][0] = ord(lc)
  if asciiA <= ord(rc) <= asciiZ:
    childArr[ord(parent)][1] = ord(rc)

output(f"{VLR(asciiA)}\n")
output(f"{LVR(asciiA)}\n")
output(f"{LRV(asciiA)}\n")



  

File: 1. 단계별로_풀어보기/33. 트리/4. 1991.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const treeInfoArr = inputArr.slice(1).map(row => row.trim().split(' '))

// 자식노드들의 목록 배열
const childArr = Array.from({length: 256}, () => [null, null])

const asciiA = 'A'.charCodeAt(0)
const asciiZ = 'Z'.charCodeAt(0)

treeInfoArr.forEach(info => {
  const [parent, lchild, rchild] = info

  const asciiParent = parent.charCodeAt(0)
  const asciiLeft = lchild.charCodeAt(0)
  const asciiRight = rchild.charCodeAt(0)

  if (asciiA <= asciiLeft && asciiLeft <= asciiZ) {
    childArr[asciiParent][0] = asciiLeft
  }
  if (asciiA <= asciiRight && asciiRight <= asciiZ) {
    childArr[asciiParent][1] = asciiRight
  }
})

process.stdout.write(`${VLR(asciiA)}\n`)
process.stdout.write(`${LVR(asciiA)}\n`)
process.stdout.write(`${LRV(asciiA)}\n`)

function VLR(now) {
  let result = String.fromCharCode(now)

  let lchild = childArr[now][0]
  let rchild = childArr[now][1]

  if (lchild) {
    result += VLR(lchild)
  }
  if (rchild) {
    result += VLR(rchild)
  }
  return result
}
function LVR(now) {
  let result = String.fromCharCode(now)

  let lchild = childArr[now][0]
  let rchild = childArr[now][1]

  if (lchild) {
    result = LVR(lchild) + result
  }
  if (rchild) {
    result = result + LVR(rchild)
  }
  return result
}
function LRV(now) {
  let result = ''

  let lchild = childArr[now][0]
  let rchild = childArr[now][1]

  if (lchild) {
    result += LRV(lchild)
  }
  if (rchild) {
    result += LRV(rchild)
  }
  result += String.fromCharCode(now)
  return result
}
